forge-game/src/main/java/forge/game/trigger/TriggerSpellAbilityCastOrCopy.java

```diff
@@ -111,6 +111,24 @@ public final boolean performTest(final Map<AbilityKey, Object> runParams) {
                     return false;
                 }
             }
+            if (hasParam("ActivatorThisTurnCastEach")) {
+                final String compare = getParam("ActivatorThisTurnCastEach");
+                final String valid = getParamOrDefault("ValidCard", "Card");
+                boolean found = false;
+                int right = Integer.parseInt(compare.substring(2));
+                for (String v : valid.split(",")) {
+                    List<Card> thisTurnCast = CardUtil.getThisTurnCast(v, getHostCard(), this, getHostCard().getController());
+                    thisTurnCast = CardLists.filterControlledByAsList(thisTurnCast, activator);
+                    int left = thisTurnCast.size();
+                    if (!Expressions.compare(left, compare, right)) {
```

Comment:
this check shouldn't be negated now?

Imo more logic is needed: you can't check against all valid types here again but only those that the triggered spell actually matches

Reply:
done?